From 7ab8b790a75cbb296e2d30f293a804d3c621e162 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Thu, 8 Oct 2026 15:59:36 +0200
Subject: [PATCH] Comments of the generic-parallel interfaces updated,
 preexisting documentation errors fixed

Memory space of the batches, sycl queues (required, in-order for the
asynchronous computations), lists of backends, conventions of the
backends of the material properties, generated documentation (strides,
parameters of the constructor, order of the arguments of the initialize
functions). Preexisting errors in the touched files: misplaced comments,
wrong parameter names, template parameters documented as parameters.
---
 include/TFEL/FSAlgorithm/accumulate.hxx       | 18 ++++++++----
 include/TFEL/FSAlgorithm/equal.hxx            | 18 ++++++++----
 include/TFEL/FSAlgorithm/fill.hxx             |  3 +-
 include/TFEL/FSAlgorithm/for_each.hxx         | 14 +++++-----
 include/TFEL/FSAlgorithm/generate.hxx         | 10 ++++---
 include/TFEL/FSAlgorithm/max_element.hxx      | 22 ++++++++++-----
 include/TFEL/FSAlgorithm/min_element.hxx      | 24 ++++++++++------
 include/TFEL/FSAlgorithm/swap_ranges.hxx      | 10 ++++---
 include/TFEL/FSAlgorithm/transform.hxx        | 27 +++++++++++-------
 include/TFEL/Material/BoundsCheck.hxx         | 10 +++----
 .../Math/ST2toST2/ST2toST2TransposeExpr.hxx   |  7 ++---
 .../Stensor/Internals/SortEigenValues.hxx     |  2 +-
 .../Stensor/Internals/SortEigenVectors.hxx    |  2 +-
 .../LeftCauchyGreenTensorDerivativeExpr.hxx   |  6 ++--
 .../RightCauchyGreenTensorDerivativeExpr.hxx  |  6 ++--
 .../MFront/GenericBehaviour/Integrate.hxx     |  2 +-
 .../MaterialProperty/AbstractBackend.hxx      |  2 +-
 .../MaterialProperty/BackendBase.hxx          | 14 ++++++----
 .../MaterialProperty/CUDABackendBase.hxx      | 23 +++++++++------
 .../CxxProgrammingModelBackendBase.hxx        | 28 +++++++++++--------
 .../MaterialProperty/KokkosBackend.hxx        |  2 +-
 .../MaterialProperty/OpenMPBackend.hxx        |  2 +-
 .../MaterialProperty/ParallelSTLBackend.hxx   |  2 +-
 .../MaterialProperty/SYCLBackend.hxx          |  2 +-
 .../GenericParallelBehaviour/BatchData.h      | 16 ++++++++---
 .../GenericParallelBehaviourInterface.hxx     |  5 ++--
 .../MaterialProperty.h                        |  2 ++
 mfront/src/GenericBehaviourInterface.cxx      | 18 ++++++++----
 .../MaterialProperty/BackendBase.cxx          |  7 +++--
 .../PolyCrystalSharedDataTest.mfront          |  2 +-
 .../generic-parallel/test-cxx.cxx             |  6 ++--
 31 files changed, 192 insertions(+), 120 deletions(-)

diff --git a/include/TFEL/FSAlgorithm/accumulate.hxx b/include/TFEL/FSAlgorithm/accumulate.hxx
index bb1aafd08..4205aa68b 100644
--- a/include/TFEL/FSAlgorithm/accumulate.hxx
+++ b/include/TFEL/FSAlgorithm/accumulate.hxx
@@ -21,7 +21,8 @@ namespace tfel::fsalgo {
   /*!
    * \class accumulate
    * \brief accumulate computes the sum (or some other binary operation) of N
-   * elements of a container. \param N number of element to be summed.
+   * elements of a container.
+   * \tparam N: number of elements to be summed.
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/transform.html
@@ -34,8 +35,10 @@ namespace tfel::fsalgo {
   template <unsigned int N>
   struct accumulate {
     /*!
-     * \param InputIterator iterator of the sequence to be summed
-     * \param T  type of the result
+     * \tparam InputIterator: iterator of the sequence to be summed
+     * \tparam T: type of the result
+     * \param[in] p: iterator to the first element of the sequence
+     * \param[in] init: initial value
      *
      * \return result of the summation:
      * \f[\displaystyle\mathrm{init}+\sum_{i=0}^{N-1}*(p+i)\f]
@@ -60,9 +63,12 @@ namespace tfel::fsalgo {
 
     /*!
      *
-     * \param InputIterator  iterator of the sequence to be operated on
-     * \param T              type of the result
-     * \param BinaryFunction function doing the operation
+     * \tparam InputIterator: iterator of the sequence to be operated on
+     * \tparam T: type of the result
+     * \tparam BinaryFunction: function doing the operation
+     * \param[in] p: iterator to the first element of the sequence
+     * \param[in] init: initial value
+     * \param[in] binary_op: function doing the operation
      * \return result of the operation :
      *
      \f[\displaystyle\mathrm{Op}\left(*(p+N-1),\ldots,\mathrm{Op}\left(*p,\mathrm{init}\right)\ldots\right)\f]
diff --git a/include/TFEL/FSAlgorithm/equal.hxx b/include/TFEL/FSAlgorithm/equal.hxx
index c2edf8806..6bda277b1 100644
--- a/include/TFEL/FSAlgorithm/equal.hxx
+++ b/include/TFEL/FSAlgorithm/equal.hxx
@@ -22,7 +22,8 @@ namespace tfel::fsalgo {
    * \class equal
    * \brief Equal returns true if the two ranges [first1, first1+N) and [first2,
    * first2 + N) are identical when compared element-by-element, and otherwise
-   * returns false \param N number of element to be compared.
+   * returns false.
+   * \tparam N: number of elements to be compared.
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/equal.html
@@ -35,8 +36,10 @@ namespace tfel::fsalgo {
   template <unsigned int N>
   struct equal {
     /*!
-     * \param  InputIterator  iterator for the first  sequence
-     * \param  InputIterator2 iterator for the second sequence
+     * \tparam InputIterator: iterator for the first sequence
+     * \tparam InputIterator2: iterator for the second sequence
+     * \param[in] p: iterator to the first element of the first sequence
+     * \param[in] q: iterator to the first element of the second sequence
      *
      * \return true if all elements are equal
      *
@@ -62,9 +65,12 @@ namespace tfel::fsalgo {
     }
 
     /*!
-     * \param  InputIterator  iterator for the first  sequence
-     * \param  InputIterator2 iterator for the second sequence
-     * \param  BinaryPredicate predicate used to make comparison
+     * \tparam InputIterator: iterator for the first sequence
+     * \tparam InputIterator2: iterator for the second sequence
+     * \tparam BinaryPredicate: predicate used to make comparison
+     * \param[in] p: iterator to the first element of the first sequence
+     * \param[in] q: iterator to the first element of the second sequence
+     * \param[in] binary_pred: predicate used to make comparison
      * \return true if all elements are one-by-one equal given the predicate
      * used.
      *
diff --git a/include/TFEL/FSAlgorithm/fill.hxx b/include/TFEL/FSAlgorithm/fill.hxx
index 545e2a435..992de37f9 100644
--- a/include/TFEL/FSAlgorithm/fill.hxx
+++ b/include/TFEL/FSAlgorithm/fill.hxx
@@ -22,7 +22,8 @@ namespace tfel::fsalgo {
    * \class fill
    * \brief Fill assigns the value value to every element in the range [first,
    * last). That is, for every iterator i in [first, first+N), it performs the
-   * assignment *i = value. \param N number of element to be filled.
+   * assignment *i = value.
+   * \tparam N: number of elements to be filled.
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/fill.html
diff --git a/include/TFEL/FSAlgorithm/for_each.hxx b/include/TFEL/FSAlgorithm/for_each.hxx
index 15924ca66..4256faab1 100644
--- a/include/TFEL/FSAlgorithm/for_each.hxx
+++ b/include/TFEL/FSAlgorithm/for_each.hxx
@@ -23,9 +23,9 @@ namespace tfel::fsalgo {
    * \brief For_each applies the function object f to each element
    * in the range [first, first+N); f's return value, if any, is
    * ignored. Applications are performed in forward order, i.e. from
-   * first to last. For_each returns the function object after it
-   * has been applied to each element.
-   * \param N number of element to which f will be applied
+   * first to last. The function object is passed by reference: its
+   * state after the last application is available to the caller.
+   * \tparam N: number of elements to which f will be applied
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/for_each.html
@@ -38,10 +38,10 @@ namespace tfel::fsalgo {
   template <unsigned int N>
   struct for_each {
     /*!
-     * \param InputIterator input iterator
-     * \param UnaryFunction unary function which is applied to the elements
-     *
-     * \return UnaryFunction
+     * \tparam InputIterator: input iterator
+     * \tparam UnaryFunction: unary function which is applied to the elements
+     * \param[in] p: iterator to the first element
+     * \param[in,out] f: function applied to each element
      *
      * \pre
      * - InputIterator is a model of Input Iterator
diff --git a/include/TFEL/FSAlgorithm/generate.hxx b/include/TFEL/FSAlgorithm/generate.hxx
index d6a8c998e..d06f91185 100644
--- a/include/TFEL/FSAlgorithm/generate.hxx
+++ b/include/TFEL/FSAlgorithm/generate.hxx
@@ -21,8 +21,8 @@ namespace tfel::fsalgo {
   /*!
    * \class generate
    * \brief generate assigns the result of invoking gen, a function object that
-   * takes no arguments, to each element in the range [first, last). \param N
-   * number of element to be which generate is applied.
+   * takes no arguments, to each element in the range [first, first+N).
+   * \tparam N: number of elements to which generate is applied.
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/generate.html
@@ -35,8 +35,10 @@ namespace tfel::fsalgo {
   template <unsigned int N>
   struct generate {
     /*!
-     * \param p: iterator to the elements
-     * \param g: generator type used
+     * \tparam ForwardIterator: iterator to the elements
+     * \tparam Generator: type of the generator
+     * \param[in] p: iterator to the first element
+     * \param[in] gen: generator
      */
     template <typename ForwardIterator, typename Generator>
     TFEL_HOST_DEVICE static void exe(ForwardIterator p, Generator gen) {
diff --git a/include/TFEL/FSAlgorithm/max_element.hxx b/include/TFEL/FSAlgorithm/max_element.hxx
index 6260d3fa2..2f3183cd8 100644
--- a/include/TFEL/FSAlgorithm/max_element.hxx
+++ b/include/TFEL/FSAlgorithm/max_element.hxx
@@ -21,7 +21,7 @@ namespace tfel::fsalgo {
   /*!
    * \class max_element
    * \brief max_element finds the largest element in the range [first, first+N)
-   * \param N number of element to be transformed.
+   * \tparam N: number of elements of the range.
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/max_element.html
@@ -33,7 +33,9 @@ namespace tfel::fsalgo {
   struct max_element {
     /*
      * Internal function member used by max_element<N>::exe
-     * \param ForwardIterator, iterator to the elements.
+     * \tparam ForwardIterator: iterator to the elements.
+     * \param[in] p: iterator to the next element to be compared
+     * \param[in] q: iterator to the largest element found so far
      *
      * \return ForwardIterator to the maximum element
      *
@@ -54,8 +56,11 @@ namespace tfel::fsalgo {
     /*
      * Internal function member used by max_element<N>::exe
      *
-     * \param ForwardIterator, iterator to the elements.
-     * \param BinaryPredicate, comparison function
+     * \tparam ForwardIterator: iterator to the elements.
+     * \tparam BinaryPredicate: comparison function
+     * \param[in] p: iterator to the next element to be compared
+     * \param[in] q: iterator to the element selected so far
+     * \param[in] comp: comparison function
      *
      * \return ForwardIterator to the maximum element
      *
@@ -81,7 +86,8 @@ namespace tfel::fsalgo {
      * iterator in [first, last) points to a value greater than *i. The return
      * value is first+N if and only if [first, last) is an empty range.
      *
-     * \param ForwardIterator, iterator to the elements.
+     * \tparam ForwardIterator: iterator to the elements.
+     * \param[in] p: iterator to the first element
      *
      * \return ForwardIterator to the maximum element
      *
@@ -104,8 +110,10 @@ namespace tfel::fsalgo {
      * such that, for every iterator j in [first, first+N), comp(*i, *j) is
      * false, where comp is a binary predicate given in argument
      *
-     * \param ForwardIterator, iterator to the elements.
-     * \param BinaryPredicate, comparison function
+     * \tparam ForwardIterator: iterator to the elements.
+     * \tparam BinaryPredicate: comparison function
+     * \param[in] p: iterator to the first element
+     * \param[in] comp: comparison function
      *
      * \return ForwardIterator to the maximum element
      *
diff --git a/include/TFEL/FSAlgorithm/min_element.hxx b/include/TFEL/FSAlgorithm/min_element.hxx
index 2c653138e..28fd6271f 100644
--- a/include/TFEL/FSAlgorithm/min_element.hxx
+++ b/include/TFEL/FSAlgorithm/min_element.hxx
@@ -21,7 +21,7 @@ namespace tfel::fsalgo {
   /*!
    * \class min_element
    * \brief min_element finds the smallest element in the range [first, first+N)
-   * \param N number of element to be transformed.
+   * \tparam N: number of elements of the range.
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/min_element.html
@@ -33,7 +33,9 @@ namespace tfel::fsalgo {
   struct min_element {
     /*
      * Internal function member used by min_element<N>::exe
-     * \param ForwardIterator, iterator to the elements.
+     * \tparam ForwardIterator: iterator to the elements.
+     * \param[in] p: iterator to the next element to be compared
+     * \param[in] q: iterator to the smallest element found so far
      *
      * \return ForwardIterator to the minimum element
      *
@@ -59,8 +61,11 @@ namespace tfel::fsalgo {
     /*
      * Internal function member used by min_element<N>::exe
      *
-     * \param ForwardIterator, iterator to the elements.
-     * \param BinaryPredicate, comparison function
+     * \tparam ForwardIterator: iterator to the elements.
+     * \tparam BinaryPredicate: comparison function
+     * \param[in] p: iterator to the next element to be compared
+     * \param[in] q: iterator to the element selected so far
+     * \param[in] comp: comparison function
      *
      * \return ForwardIterator to the minimum element
      *
@@ -88,10 +93,11 @@ namespace tfel::fsalgo {
    public:
     /*!
      * It returns the first iterator i in [first, first+N) such that no other
-     * iterator in [first, last) points to a value greater than *i. The return
+     * iterator in [first, last) points to a value smaller than *i. The return
      * value is first+N if and only if [first, last) is an empty range.
      *
-     * \param ForwardIterator, iterator to the elements.
+     * \tparam ForwardIterator: iterator to the elements.
+     * \param[in] p: iterator to the first element
      *
      * \return ForwardIterator to the minimum element
      *
@@ -114,8 +120,10 @@ namespace tfel::fsalgo {
      * such that, for every iterator j in [first, first+N), comp(*i, *j) is
      * false, where comp is a binary predicate given in argument
      *
-     * \param ForwardIterator, iterator to the elements.
-     * \param BinaryPredicate, comparison function
+     * \tparam ForwardIterator: iterator to the elements.
+     * \tparam BinaryPredicate: comparison function
+     * \param[in] p: iterator to the first element
+     * \param[in] comp: comparison function
      *
      * \return ForwardIterator to the minimum element
      *
diff --git a/include/TFEL/FSAlgorithm/swap_ranges.hxx b/include/TFEL/FSAlgorithm/swap_ranges.hxx
index 0dbadf43e..facd17bd6 100644
--- a/include/TFEL/FSAlgorithm/swap_ranges.hxx
+++ b/include/TFEL/FSAlgorithm/swap_ranges.hxx
@@ -24,8 +24,8 @@ namespace tfel::fsalgo {
    * \brief swap_ranges swaps each of the elements in the range [first1,
    * first1+N) with the corresponding element in the range [first2, first2 + N).
    * That is, for each integer n such that 0 <= i < N, it swaps *(first1 + i)
-   * and *(first2 + i). The return value is first2 + N. \param N size of the
-   * ranges to be swapped.
+   * and *(first2 + i). The return value is first2 + N.
+   * \tparam N: size of the ranges to be swapped.
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/swap_ranges.html
@@ -36,8 +36,10 @@ namespace tfel::fsalgo {
   template <unsigned int N>
   struct swap_ranges {
     /*!
-     * \param ForwardIterator1
-     * \param ForwardIterator
+     * \tparam ForwardIterator1: iterator to the first range
+     * \tparam ForwardIterator2: iterator to the second range
+     * \param[in] p: iterator to the first element of the first range
+     * \param[in] q: iterator to the first element of the second range
      *
      * \return  q + N
      *
diff --git a/include/TFEL/FSAlgorithm/transform.hxx b/include/TFEL/FSAlgorithm/transform.hxx
index a8253cc13..621f49722 100644
--- a/include/TFEL/FSAlgorithm/transform.hxx
+++ b/include/TFEL/FSAlgorithm/transform.hxx
@@ -21,7 +21,7 @@ namespace tfel::fsalgo {
   /*!
    * \class transform
    * \brief transform performs an operation on objects
-   * \param N number of element to be transformed.
+   * \tparam N: number of elements to be transformed.
    *
    * This documentation is mostly inspired from:
    * http://www.sgi.com/tech/stl/transform.html
@@ -39,9 +39,12 @@ namespace tfel::fsalgo {
      * that operation to *o, where o is the corresponding output iterator. The
      * return value is result + N.
      *
-     * \param  InputIterator  iterator to the element to be transformed.
-     * \param  OutputIterator iterator to which result is placed.
-     * \param  UnaryFunction  function used to perform transformation.
+     * \tparam InputIterator: iterator to the elements to be transformed.
+     * \tparam OutputIterator: iterator to which the results are placed.
+     * \tparam UnaryFunction: function used to perform the transformation.
+     * \param[in] p: iterator to the first element to be transformed
+     * \param[in] q: iterator to which the first result is placed
+     * \param[in] op: function used to perform the transformation
      *
      * \return OutputIterator iterator after the last element transformed.
      *
@@ -76,10 +79,14 @@ namespace tfel::fsalgo {
      * is, for each n such that 0 <= i < N, it performs the assignment *(result
      * + i) = op(*(first1 + i), *(first2 + i). The return value is result + N.
      *
-     * \param  InputIterator1  iterator to the first  element to be transformed.
-     * \param  InputIterator2  iterator to the second element to be transformed.
-     * \param  OutputIterator iterator to which result is placed.
-     * \param  UnaryFunction  function used to perform transformation
+     * \tparam InputIterator1: iterator to the first input range.
+     * \tparam InputIterator2: iterator to the second input range.
+     * \tparam OutputIterator: iterator to which the results are placed.
+     * \tparam BinaryFunction: function used to perform the transformation
+     * \param[in] p: iterator to the first element of the first input range
+     * \param[in] q: iterator to the first element of the second input range
+     * \param[in] r: iterator to which the first result is placed
+     * \param[in] op: function used to perform the transformation
      *
      * \return OutputIterator iterator after the last element transformed.
      *
@@ -118,7 +125,7 @@ namespace tfel::fsalgo {
   template <>
   struct transform<0u> {
     /*!
-     * \return the ouptut iterator to the end the recursion
+     * \return the output iterator, to end the recursion
      * \sa transform<N>::exe()
      *
      * \author Thomas Helfer
@@ -134,7 +141,7 @@ namespace tfel::fsalgo {
     }
 
     /*!
-     * \return the ouptut iterator to the end the recursion
+     * \return the output iterator, to end the recursion
      * \sa transform<N>::exe()
      *
      * \author Thomas Helfer
diff --git a/include/TFEL/Material/BoundsCheck.hxx b/include/TFEL/Material/BoundsCheck.hxx
index 82d9af01c..10c6d9127 100644
--- a/include/TFEL/Material/BoundsCheck.hxx
+++ b/include/TFEL/Material/BoundsCheck.hxx
@@ -39,19 +39,19 @@ namespace std {
 namespace tfel::material {
 
   /*!
-   * \brief an helper class
+   * \brief a helper class
    *
    * All the checks return a boolean which is false if a bound is violated
    * and if this violation is not tolerated by the out of bounds policy.
    *
-   * The names of the variables are passed as C strings: constructing a
-   * `std::string_view` from a C string is not supported in device code by
-   * some compilers (nvcc silently removes the code).
-   *
    * On the host, an exception is thrown in this case, so the returned value
    * is always true. In device code, where exceptions are not available, the
    * returned value must be used to report the failure. Warnings are not
    * displayed in device code.
+   *
+   * The names of the variables are passed as C strings: constructing a
+   * `std::string_view` from a C string is not supported in device code by
+   * some compilers (nvcc silently removes the code).
    */
   struct BoundsCheckBase {
     /*!
diff --git a/include/TFEL/Math/ST2toST2/ST2toST2TransposeExpr.hxx b/include/TFEL/Math/ST2toST2/ST2toST2TransposeExpr.hxx
index 158ef8efd..92b5a49e0 100644
--- a/include/TFEL/Math/ST2toST2/ST2toST2TransposeExpr.hxx
+++ b/include/TFEL/Math/ST2toST2/ST2toST2TransposeExpr.hxx
@@ -53,16 +53,15 @@ namespace tfel::math {
     typedef IndexType size_type;
     typedef ptrdiff_t difference_type;
     /*!
-     * \brief access operator
-     * \param[in] i: row number
-     * \param[in] j: colum number
+     * \brief constructor
+     * \param[in] l: transposed object
      */
     TFEL_HOST_DEVICE constexpr explicit ST2toST2TransposeExpr(A l) noexcept
         : a(l) {}
     /*!
      * \brief access operator
      * \param[in] i: row number
-     * \param[in] j: colum number
+     * \param[in] j: column number
      */
     [[nodiscard]] TFEL_HOST_DEVICE constexpr NumType operator()(
         const IndexType i, const IndexType j) const noexcept {
diff --git a/include/TFEL/Math/Stensor/Internals/SortEigenValues.hxx b/include/TFEL/Math/Stensor/Internals/SortEigenValues.hxx
index 517a058c4..677dc6bf5 100644
--- a/include/TFEL/Math/Stensor/Internals/SortEigenValues.hxx
+++ b/include/TFEL/Math/Stensor/Internals/SortEigenValues.hxx
@@ -40,7 +40,7 @@ namespace tfel::math::internals {
     TFEL_HOST_DEVICE static void exe(T&, T&, T&, const EigenValuesOrdering) {
     }  // end of exe
   };  // end of struct SortEigenValues<1u>
-  //! \brief partial specialisation of the `SortEigenValues` in 1D
+  //! \brief partial specialisation of the `SortEigenValues` in 2D
   template <>
   struct SortEigenValues<2u> {
     //! a simple alias
diff --git a/include/TFEL/Math/Stensor/Internals/SortEigenVectors.hxx b/include/TFEL/Math/Stensor/Internals/SortEigenVectors.hxx
index ab20826b0..f968c47ca 100644
--- a/include/TFEL/Math/Stensor/Internals/SortEigenVectors.hxx
+++ b/include/TFEL/Math/Stensor/Internals/SortEigenVectors.hxx
@@ -42,7 +42,7 @@ namespace tfel::math::internals {
                                      const EigenValuesOrdering) {
     }  // end of exe
   };  // end of struct SortEigenVectors<1u>
-  //! \brief partial specialisation of the `SortEigenVectors` in 1D
+  //! \brief partial specialisation of the `SortEigenVectors` in 2D
   template <>
   struct SortEigenVectors<2u> {
     //! a simple alias
diff --git a/include/TFEL/Math/T2toST2/LeftCauchyGreenTensorDerivativeExpr.hxx b/include/TFEL/Math/T2toST2/LeftCauchyGreenTensorDerivativeExpr.hxx
index e7b599cbf..2e29221b6 100644
--- a/include/TFEL/Math/T2toST2/LeftCauchyGreenTensorDerivativeExpr.hxx
+++ b/include/TFEL/Math/T2toST2/LeftCauchyGreenTensorDerivativeExpr.hxx
@@ -45,7 +45,7 @@ namespace tfel::math {
     typedef numeric_type<T2toST2Type> value_type;
     // NOLINTBEGIN(google-explicit-constructor)
     /*!
-     * \param[in] B : second tensor of the product
+     * \param[in] F: deformation gradient
      */
     template <TensorConcept TensorType>
     TFEL_HOST_DEVICE constexpr Expr(const TensorType& F) noexcept {
@@ -89,7 +89,7 @@ namespace tfel::math {
     typedef numeric_type<T2toST2Type> value_type;
     // NOLINTBEGIN(google-explicit-constructor)
     /*!
-     * \param[in] B : second tensor of the product
+     * \param[in] F: deformation gradient
      */
     template <TensorConcept TensorType>
     TFEL_HOST_DEVICE constexpr Expr(const TensorType& F) noexcept {
@@ -145,7 +145,7 @@ namespace tfel::math {
     typedef numeric_type<T2toST2Type> value_type;
     // NOLINTBEGIN(google-explicit-constructor)
     /*!
-     * \param[in] B : second tensor of the product
+     * \param[in] F: deformation gradient
      */
     template <TensorConcept TensorType>
     TFEL_HOST_DEVICE constexpr Expr(const TensorType& F) noexcept {
diff --git a/include/TFEL/Math/T2toST2/RightCauchyGreenTensorDerivativeExpr.hxx b/include/TFEL/Math/T2toST2/RightCauchyGreenTensorDerivativeExpr.hxx
index f2990b51e..7fc31b838 100644
--- a/include/TFEL/Math/T2toST2/RightCauchyGreenTensorDerivativeExpr.hxx
+++ b/include/TFEL/Math/T2toST2/RightCauchyGreenTensorDerivativeExpr.hxx
@@ -45,7 +45,7 @@ namespace tfel::math {
     typedef numeric_type<T2toST2Type> value_type;
     // NOLINTBEGIN(google-explicit-constructor)
     /*!
-     * \param[in] B : second tensor of the product
+     * \param[in] F: deformation gradient
      */
     template <TensorConcept TensorType>
     TFEL_HOST_DEVICE constexpr Expr(const TensorType& F) noexcept {
@@ -95,7 +95,7 @@ namespace tfel::math {
     typedef numeric_type<T2toST2Type> value_type;
     // NOLINTBEGIN(google-explicit-constructor)
     /*!
-     * \param[in] B : second tensor of the product
+     * \param[in] F: deformation gradient
      */
     template <TensorConcept TensorType>
     TFEL_HOST_DEVICE constexpr Expr(const TensorType& F) noexcept {
@@ -156,7 +156,7 @@ namespace tfel::math {
     typedef numeric_type<T2toST2Type> value_type;
     // NOLINTBEGIN(google-explicit-constructor)
     /*!
-     * \param[in] B : second tensor of the product
+     * \param[in] F: deformation gradient
      */
     template <TensorConcept TensorType>
     TFEL_HOST_DEVICE constexpr Expr(const TensorType& F) noexcept {
diff --git a/mfront/include/MFront/GenericBehaviour/Integrate.hxx b/mfront/include/MFront/GenericBehaviour/Integrate.hxx
index 3b456b580..d40aa5190 100644
--- a/mfront/include/MFront/GenericBehaviour/Integrate.hxx
+++ b/mfront/include/MFront/GenericBehaviour/Integrate.hxx
@@ -392,7 +392,7 @@ namespace mfront::gb {
    * \param[in,out] b: behaviour
    * \param[in,out] d: behaviour data
    * \param[in] o: computation options
-   * \param[in,out] smflag: consistent tangent operator expected
+   * \param[in] f: choice of consistent tangent operator
    */
   template <typename Behaviour>
   TFEL_HOST_DEVICE int computePredictionOperator(
diff --git a/mfront/include/MFront/GenericParallel/MaterialProperty/AbstractBackend.hxx b/mfront/include/MFront/GenericParallel/MaterialProperty/AbstractBackend.hxx
index 9c6e9bad4..99e58e080 100644
--- a/mfront/include/MFront/GenericParallel/MaterialProperty/AbstractBackend.hxx
+++ b/mfront/include/MFront/GenericParallel/MaterialProperty/AbstractBackend.hxx
@@ -34,7 +34,7 @@ namespace mfront::generic_parallel::material_property {
 
   /*!
    * \brief an abstract class which aims at writing code for a specific backend
-   * (CUDA, SYCL, parallel STL)
+   * (CUDA, HIP, Kokkos, OpenMP, parallel STL, SYCL)
    */
   struct MFRONT_VISIBILITY_EXPORT AbstractBackend {
     //! \brief return the name of the backend
diff --git a/mfront/include/MFront/GenericParallel/MaterialProperty/BackendBase.hxx b/mfront/include/MFront/GenericParallel/MaterialProperty/BackendBase.hxx
index 237b98ba7..36bb0180d 100644
--- a/mfront/include/MFront/GenericParallel/MaterialProperty/BackendBase.hxx
+++ b/mfront/include/MFront/GenericParallel/MaterialProperty/BackendBase.hxx
@@ -241,14 +241,16 @@ namespace mfront::generic_parallel::material_property {
      * \brief return if the backend shall handle the data transfer from a host
      * to a device
      *
-     * If this method return true, the argument of the C function
-     * implementations containing pointers to values of the state variables
+     * If this method returns true, the argument of the C function
+     * implementations containing pointers to values of the arguments
      * is named `mfront_args_host`. The backend shall
-     * then defined a pointer named `mfront_args` to the beginning of an
-     * internal array of pointers.
+     * then define a pointer named `mfront_args` to the beginning of an
+     * internal array of pointers. The output argument is then named
+     * `mfront_output_host`, and the backend shall define a pointer named
+     * `mfront_output`.
      *
-     * If this method returns false, this argument is directly named
-     * `mfront_args`.
+     * If this method returns false, those arguments are directly named
+     * `mfront_args` and `mfront_output`.
      *
      * \note by default, no data transfer is handled
      */
diff --git a/mfront/include/MFront/GenericParallel/MaterialProperty/CUDABackendBase.hxx b/mfront/include/MFront/GenericParallel/MaterialProperty/CUDABackendBase.hxx
index 17d8776a8..f31977a99 100644
--- a/mfront/include/MFront/GenericParallel/MaterialProperty/CUDABackendBase.hxx
+++ b/mfront/include/MFront/GenericParallel/MaterialProperty/CUDABackendBase.hxx
@@ -23,7 +23,8 @@
 namespace mfront::generic_parallel::material_property {
 
   /*!
-   * \brief This class implements a backend for the parallel STL algorithms
+   * \brief base class of the CUDA and HIP backends: each point is evaluated
+   * by one thread of the device
    */
   struct MFRONT_VISIBILITY_EXPORT CUDABackendBase : BackendBase {
     //! \brief default constructor
@@ -60,8 +61,8 @@ namespace mfront::generic_parallel::material_property {
         const GenericParallelMaterialPropertyInterface&,
         const MaterialPropertyDescription&,
         const FileDescription&) const override;
-    //! \brief destructor
     [[nodiscard]] std::string getMemorySpace() const override;
+    //! \brief destructor
     ~CUDABackendBase() noexcept override;
 
    protected:
@@ -74,8 +75,9 @@ namespace mfront::generic_parallel::material_property {
      * \param[in] i: interface
      * \param[in] mpd: material property description
      * \param[in] fd: file description
-     * \param[in] treatUniformArguments: boolean stating if one shall add
-     * support for uniform arguments
+     * \param[in] treatStrides: if true, the kernel handles the strides of the
+     * output and of the arguments. Otherwise, the name of the kernel ends
+     * with `_without_strides`.
      */
     virtual void writeGlobalFunctions2(
         std::ostream&,
@@ -90,8 +92,10 @@ namespace mfront::generic_parallel::material_property {
      * \param[in] i: interface
      * \param[in] mpd: material property description
      * \param[in] fd: file description
-     * \param[in] treatUniformArguments: boolean stating if one shall add
-     * support for uniform arguments
+     * \param[in] treatStrides: if true, the function handles the strides of
+     * the output and of the arguments. Otherwise, the arguments and the
+     * output are contiguous arrays and the name of the function ends with
+     * `2`.
      */
     virtual void writeCImplementations2(
         std::ostream&,
@@ -100,13 +104,14 @@ namespace mfront::generic_parallel::material_property {
         const FileDescription&,
         const bool) const;
     /*!
-     * \brief write the call the kernel inside the body of the material property
+     * \brief write the call to the kernel inside the body of the material
+     * property
      *
      * \param[in] os: output stream
      * \param[in] i: interface
      * \param[in] mpd: material property description
-     * \param[in] treatUniformArguments: boolean stating if one shall add
-     * support for uniform arguments
+     * \param[in] treatStrides: boolean stating if strides are treated (see
+     * `writeKernelSelection`)
      */
     virtual void writeKernelCall(
         std::ostream&,
diff --git a/mfront/include/MFront/GenericParallel/MaterialProperty/CxxProgrammingModelBackendBase.hxx b/mfront/include/MFront/GenericParallel/MaterialProperty/CxxProgrammingModelBackendBase.hxx
index 82bcc041e..b319797ff 100644
--- a/mfront/include/MFront/GenericParallel/MaterialProperty/CxxProgrammingModelBackendBase.hxx
+++ b/mfront/include/MFront/GenericParallel/MaterialProperty/CxxProgrammingModelBackendBase.hxx
@@ -24,8 +24,8 @@ namespace mfront::generic_parallel::material_property {
 
   /*!
    * \brief This intermediate class helps implementing various closely related
-   * backends which defines the kernel as a lambda function. Those backends
-   * include `ParallelSTL`, `SYCL`
+   * backends which define the kernel as a lambda function. Those backends
+   * include `ParallelSTL`, `SYCL`, `Kokkos` and `OpenMP`
    */
   struct MFRONT_VISIBILITY_EXPORT CxxProgrammingModelBackendBase : BackendBase {
     //! \brief default constructor
@@ -76,7 +76,7 @@ namespace mfront::generic_parallel::material_property {
      * \param[in] mpd: material property description
      * \param[in] fd: file description
      * \param[in] handleStrides: boolean stating if one shall add
-     * support for striedes
+     * support for strides
      */
     virtual void writeCImplementations2(
         std::ostream&,
@@ -102,9 +102,13 @@ namespace mfront::generic_parallel::material_property {
                              const bool) const;
     /*!
      * \brief write the declarations of the variables captured by the kernels
-     * (pointer to the bounds statuses, stored in a container named
-     * `mfront_bounds_statuses_storage` declared by the backend, and values of
-     * the parameters) and the kernels (with and without strides)
+     * (copies of the pointers to the arguments, named `mfront_arg0`,
+     * `mfront_arg1`, etc., and values of the parameters) and the kernels
+     * (without strides and, if strides are treated, with strides).
+     *
+     * If bounds checking is required, the array of statuses
+     * `mfront_bounds_statuses` must have been declared before by the backend
+     * (see `writeBoundsStatusesDeclaration`).
      *
      * \param[in] os: output stream
      * \param[in] i: interface
@@ -147,15 +151,17 @@ namespace mfront::generic_parallel::material_property {
     /*!
      * \brief write the code transferring the data to the device
      *
-     * This method is only called if the `handlesDataTransfer` returns true.
-     * This method shall define a pointer named `mfront_arg`.
-     * All memory allocated on the device shall be handled by RAII
+     * This method is only called if the `handlesDataTransfer` method returns
+     * true. This method shall define `mfront_args`, an array of pointers to
+     * the values of the arguments accessible by the device, and
+     * `mfront_output`, a pointer to the output values accessible by the
+     * device. All memory allocated on the device shall be handled by RAII.
      *
      * \param[in] os: output stream
      * \param[in] i: interface
      * \param[in] mpd: material property description
      * \param[in] handleStrides: boolean stating if one shall add
-     * support for striedes
+     * support for strides
      */
     virtual void writeDataTransfersToDevice(
         std::ostream&,
@@ -169,7 +175,7 @@ namespace mfront::generic_parallel::material_property {
      * \param[in] i: interface
      * \param[in] mpd: material property description
      * \param[in] handleStrides: boolean stating if one shall add
-     * support for striedes
+     * support for strides
      */
     virtual void writeDataTransfersToHost(
         std::ostream&,
diff --git a/mfront/include/MFront/GenericParallel/MaterialProperty/KokkosBackend.hxx b/mfront/include/MFront/GenericParallel/MaterialProperty/KokkosBackend.hxx
index c09c479d6..534d717c7 100644
--- a/mfront/include/MFront/GenericParallel/MaterialProperty/KokkosBackend.hxx
+++ b/mfront/include/MFront/GenericParallel/MaterialProperty/KokkosBackend.hxx
@@ -47,8 +47,8 @@ namespace mfront::generic_parallel::material_property {
         std::ostream&,
         const GenericParallelMaterialPropertyInterface&,
         const MaterialPropertyDescription&) const override;
-    //! \brief destructor
     [[nodiscard]] std::string getMemorySpace() const override;
+    //! \brief destructor
     ~KokkosBackend() noexcept override;
 
    protected:
diff --git a/mfront/include/MFront/GenericParallel/MaterialProperty/OpenMPBackend.hxx b/mfront/include/MFront/GenericParallel/MaterialProperty/OpenMPBackend.hxx
index cdde29b2a..c75de0fba 100644
--- a/mfront/include/MFront/GenericParallel/MaterialProperty/OpenMPBackend.hxx
+++ b/mfront/include/MFront/GenericParallel/MaterialProperty/OpenMPBackend.hxx
@@ -50,8 +50,8 @@ namespace mfront::generic_parallel::material_property {
         std::ostream&,
         const GenericParallelMaterialPropertyInterface&,
         const MaterialPropertyDescription&) const override;
-    //! \brief destructor
     [[nodiscard]] std::string getMemorySpace() const override;
+    //! \brief destructor
     ~OpenMPBackend() noexcept override;
 
    protected:
diff --git a/mfront/include/MFront/GenericParallel/MaterialProperty/ParallelSTLBackend.hxx b/mfront/include/MFront/GenericParallel/MaterialProperty/ParallelSTLBackend.hxx
index d6a1dbcac..c573da443 100644
--- a/mfront/include/MFront/GenericParallel/MaterialProperty/ParallelSTLBackend.hxx
+++ b/mfront/include/MFront/GenericParallel/MaterialProperty/ParallelSTLBackend.hxx
@@ -40,8 +40,8 @@ namespace mfront::generic_parallel::material_property {
         std::ostream&,
         const GenericParallelMaterialPropertyInterface&,
         const MaterialPropertyDescription&) const override;
-    //! \brief destructor
     [[nodiscard]] std::string getMemorySpace() const override;
+    //! \brief destructor
     ~ParallelSTLBackend() noexcept override;
 
    protected:
diff --git a/mfront/include/MFront/GenericParallel/MaterialProperty/SYCLBackend.hxx b/mfront/include/MFront/GenericParallel/MaterialProperty/SYCLBackend.hxx
index cf7584639..5049f9698 100644
--- a/mfront/include/MFront/GenericParallel/MaterialProperty/SYCLBackend.hxx
+++ b/mfront/include/MFront/GenericParallel/MaterialProperty/SYCLBackend.hxx
@@ -37,8 +37,8 @@ namespace mfront::generic_parallel::material_property {
         std::ostream&,
         const GenericParallelMaterialPropertyInterface&,
         const MaterialPropertyDescription&) const override;
-    //! \brief destructor
     [[nodiscard]] std::string getMemorySpace() const override;
+    //! \brief destructor
     ~SYCLBackend() noexcept override;
 
    protected:
diff --git a/mfront/include/MFront/GenericParallelBehaviour/BatchData.h b/mfront/include/MFront/GenericParallelBehaviour/BatchData.h
index ea2969d92..b47379e77 100644
--- a/mfront/include/MFront/GenericParallelBehaviour/BatchData.h
+++ b/mfront/include/MFront/GenericParallelBehaviour/BatchData.h
@@ -66,7 +66,8 @@ typedef struct mfront_gpb_ExecutionOptions {
    * \brief queue on which the computations are submitted. Its meaning
    * depends on the backend (`cudaStream_t` or `hipStream_t`, pointer to a
    * `sycl::queue` or to an instance of a Kokkos execution space). A null
-   * pointer denotes the default queue, if meaningful for the backend.
+   * pointer denotes the default queue, if meaningful for the backend: a
+   * queue must be given to the `sycl` backend (invalid arguments otherwise).
    */
   void* queue;
   /*!
@@ -82,7 +83,9 @@ typedef struct mfront_gpb_ExecutionOptions {
    * SYCL backend if no field is transferred): the function returns once
    * they are submitted to the queue, and the caller must synchronize the
    * queue before reading the outputs and the statistics. The output status
-   * only reports invalid arguments and runtime errors.
+   * only reports invalid arguments and runtime errors. With the `sycl`
+   * backend, the statistics require an in-order queue (invalid arguments
+   * otherwise).
    */
   mfront_gpb_BatchStatistics* statistics;
   /*!
@@ -190,8 +193,13 @@ typedef struct mfront_gpb_State {
  *
  * Integrating a batch is equivalent to calling the function generated by
  * the `generic` interface at each integration point, with the same values
- * of the time increment and of the computation options. All the pointers
- * must be accessible by the device on which the computations are performed.
+ * of the time increment and of the computation options. All the fields must
+ * be located in the memory space exported by the library (symbol
+ * `<name>_generic_parallel_memory_space`, see
+ * `MFront/GenericParallel/MemorySpace.hxx`). With the `any` memory space,
+ * the fields may be located anywhere (for instance, the `sycl` backend
+ * copies the fields which are not accessible by the device). The arrays of
+ * fields given variable by variable are only read on the host.
  */
 typedef struct mfront_gpb_BatchData {
   //! \brief number of integration points
diff --git a/mfront/include/MFront/GenericParallelBehaviourInterface.hxx b/mfront/include/MFront/GenericParallelBehaviourInterface.hxx
index 5c84cbf8f..4503f8e87 100644
--- a/mfront/include/MFront/GenericParallelBehaviourInterface.hxx
+++ b/mfront/include/MFront/GenericParallelBehaviourInterface.hxx
@@ -26,8 +26,9 @@ namespace mfront {
 
   /*!
    * \brief an interface integrating a behaviour over a batch of integration
-   * points on GPUs (CUDA, HIP) or CPUs (parallel algorithms of the C++
-   * standard library).
+   * points with a parallel programming model selected by the `backend`
+   * option (`cuda`, `hip`, `kokkos`, `openmp`, `stlpar` or `sycl`), on the
+   * host or on a device.
    *
    * Integrating a batch is equivalent to calling the function generated by
    * the `generic` interface at each integration point. This interface reuses
diff --git a/mfront/include/MFront/GenericParallelMaterialProperty/MaterialProperty.h b/mfront/include/MFront/GenericParallelMaterialProperty/MaterialProperty.h
index 73d34a5e6..93ea48d97 100644
--- a/mfront/include/MFront/GenericParallelMaterialProperty/MaterialProperty.h
+++ b/mfront/include/MFront/GenericParallelMaterialProperty/MaterialProperty.h
@@ -58,6 +58,8 @@ typedef struct mfront_gpmp_ExecutionOptions {
    * transferred), as the integration of the behaviours with statistics: the
    * function returns once the computations are submitted to the queue, and
    * the output status only reports invalid arguments and runtime errors.
+   * With the `sycl` backend, an asynchronous evaluation requires an in-order
+   * queue (invalid arguments otherwise).
    * Once the computations are completed, the
    * `<name>_generic_parallel_report` function analyses the statuses, read on
    * the host, and reports the result in an output status, as a synchronous
diff --git a/mfront/src/GenericBehaviourInterface.cxx b/mfront/src/GenericBehaviourInterface.cxx
index 0728568f2..6d997f320 100644
--- a/mfront/src/GenericBehaviourInterface.cxx
+++ b/mfront/src/GenericBehaviourInterface.cxx
@@ -300,8 +300,8 @@ namespace mfront {
       // initialize function
       for (const auto& p : d.getInitializeFunctions()) {
         out << "/*!\n"
+            << " * \\param[in,out] d: material data\n"
             << " * \\param[in] values: inputs of the initialize function\n"
-            << " * \\param[in] d: material data\n"
             << " */\n"
             << "MFRONT_SHAREDOBJ int " << f << "_InitializeFunction_" << p.first
             << "(mfront_gb_BehaviourData* const,\n"
@@ -892,15 +892,23 @@ namespace mfront {
   void GenericBehaviourInterface::writeBehaviourConstructorHeader(
       std::ostream& os,
       const BehaviourDescription& bd,
-      const Hypothesis,
+      const Hypothesis h,
       const std::string& initStateVarsIncrements) const {
     const auto tdecl = this->getBehaviourConstructorTemplateDeclaration();
     const auto dargs = this->getDataConstructorsArgumentsValues();
-    os << "/*\n"
+    // if not empty, the source of the parameters is an additional argument
+    // of the constructor (`mgpb_parameters` for the `generic-parallel`
+    // interface, see `getBehaviourConstructorArguments`)
+    const auto psrc = this->getParametersSourceInBehaviourConstructor(bd, h);
+    os << "/*!\n"
        << " * \\brief constructor for the " << this->getInterfaceName()
        << " interface\n"
-       << " * \\param[in] mgb_d: behaviour data\n"
-       << " */\n";
+       << " * \\param[in] mgb_d: behaviour data\n";
+    if (!psrc.empty()) {
+      os << " * \\param[in] " << psrc
+         << ": values of the parameters (and copies of the shared data)\n";
+    }
+    os << " */\n";
     if (!tdecl.empty()) {
       os << tdecl << '\n';
     }
diff --git a/mfront/src/GenericParallel/MaterialProperty/BackendBase.cxx b/mfront/src/GenericParallel/MaterialProperty/BackendBase.cxx
index 61d738394..92aa91c46 100644
--- a/mfront/src/GenericParallel/MaterialProperty/BackendBase.cxx
+++ b/mfront/src/GenericParallel/MaterialProperty/BackendBase.cxx
@@ -1,6 +1,6 @@
 /*!
  * \file   BackendBase.cxx
- * \brief  This file implements the `Backend` class
+ * \brief  This file implements the `BackendBase` class
  * \author Thomas Helfer
  * \date   07/04/2026
  * \copyright Copyright (C) 2006-2025 CEA/DEN, EDF R&D. All rights
@@ -169,8 +169,9 @@ namespace mfront::generic_parallel::material_property {
        << "stride of each argument. If zero, the argument is uniform\n";
     document_last_arguments();
     os << " *\n"
-       << " * \\note `mfront_npoints` is also the size of the arrays giving\n"
-       << " *   the output and the inputs when non uniform\n"
+       << " * \\note the array giving the output (resp. an argument) holds at\n"
+       << " *   least `(mfront_npoints - 1) * stride + 1` values, `stride`\n"
+       << " *   being the output stride (resp. the stride of the argument)\n"
        << " */\n"
        << "MFRONT_SHAREDOBJ void " << name << "(\n"
        << types.output_status_type << "* const,\n"
diff --git a/test-suite/behaviours/generic-parallel/PolyCrystalSharedDataTest.mfront b/test-suite/behaviours/generic-parallel/PolyCrystalSharedDataTest.mfront
index b4fb2558a..14d722f1b 100644
--- a/test-suite/behaviours/generic-parallel/PolyCrystalSharedDataTest.mfront
+++ b/test-suite/behaviours/generic-parallel/PolyCrystalSharedDataTest.mfront
@@ -30,7 +30,7 @@
 @StateVariable StrainStensor epsp;
 @StateVariable StrainStensor epsg[Np];
 
-// comment declarer 12*ng ?
+// one value per slip system and per grain
 @StateVariable strain g[Nss * Np];
 @StateVariable strain a[Nss * Np];
 @StateVariable strain p[Nss * Np];
diff --git a/test-suite/material-properties/generic-parallel/test-cxx.cxx b/test-suite/material-properties/generic-parallel/test-cxx.cxx
index 980bd4739..61bcc8444 100644
--- a/test-suite/material-properties/generic-parallel/test-cxx.cxx
+++ b/test-suite/material-properties/generic-parallel/test-cxx.cxx
@@ -394,9 +394,6 @@ struct Iconel600YoungModulusTest : public tfel::tests::TestCase {
     }
   }
 
-  // bound violated with the warning policy: the result is computed at all
-  // points and must be used with caution (status 1), as with the `generic`
-  // interface
   void test15() {
     // execution options given explicitly: instance of the execution space
     // with Kokkos, no queue otherwise
@@ -457,6 +454,9 @@ struct Iconel600YoungModulusTest : public tfel::tests::TestCase {
     }
   }
 
+  // bound violated with the warning policy: the result is computed at all
+  // points and must be used with caution (status 1), as with the `generic`
+  // interface
   void test14() {
     constexpr auto eps = double{1e-2};
     auto young = [](const double T, const double f) {
-- 
2.43.0

